blob/azureblob/azureblob.go

```diff
@@ -0,0 +1,586 @@
+		PublicAccessType: settings.PublicAccessType,
+		urls: serviceUrls{
+			serviceURL:   &serviceURL,
+			containerURL: &containerURL,
+		},
+		pageMarkers:      make(map[string]azblob.Marker),
+		defaultDelimiter: settings.DefaultDelimiter.Value,
+	}, nil
+}
+
+func openBucketWithAccountKey(ctx context.Context, settings *Settings, containerName string) (driver.Bucket, error) {
+	if settings.AccountName == "" {
+		return nil, fmt.Errorf("Settings.AccountName is not set")
+	}
+
+	if settings.AccountKey == "" {
+		return nil, fmt.Errorf("Settings.AccountKey is not set")
+	}
+
+	credential, _ := azblob.NewSharedKeyCredential(settings.AccountName, settings.AccountKey)
+	pipeline := settings.Pipeline
+	if pipeline == nil {
+		pipeline = azblob.NewPipeline(credential, azblob.PipelineOptions{})
+	}
+
+	blobURL := makeBlobStorageURL(settings.AccountName)
+	serviceURL := azblob.NewServiceURL(*blobURL, pipeline)
+	containerURL := serviceURL.NewContainerURL(containerName)
+
+	return &bucket{
+		name:             containerName,
+		credential:       credential,
+		PublicAccessType: settings.PublicAccessType,
+		urls: serviceUrls{
+			serviceURL:   &serviceURL,
+			containerURL: &containerURL,
+		},
+		pageMarkers:      make(map[string]azblob.Marker),
+		defaultDelimiter: settings.DefaultDelimiter.Value,
+	}, nil
+}
+
+func makeBlobStorageURL(accountName string) *url.URL {
+	endpoint := fmt.Sprintf("https://%s.blob.core.windows.net", accountName)
+	u, _ := url.Parse(endpoint)
+	return u
+}
+
+func (b *bucket) Delete(ctx context.Context, key string) error {
+	if key == "" {
+		return fmt.Errorf("Invalid/Empty Key")
+	}
+
+	key = strings.Replace(key, OSPathSeparator, b.defaultDelimiter, -1)
+
+	blobURL := b.urls.containerURL.NewBlockBlobURL(key)
+	_, err := blobURL.Delete(ctx, azblob.DeleteSnapshotsOptionInclude, azblob.BlobAccessConditions{})
+
+	if err != nil {
+		return makeAzureError(err, b.name, key)
+	} else {
+		return nil
+	}
+}
+
+// reader reads an azblob. It implements io.ReadCloser.
+type reader struct {
+	body  io.ReadCloser
+	attrs driver.ReaderAttributes
+	raw   *azblob.BlockBlobURL
+}
+
+func (r *reader) Read(p []byte) (int, error) {
+	return r.body.Read(p)
+}
+func (r *reader) Close() error {
+	return r.body.Close()
+}
+func (r *reader) Attributes() driver.ReaderAttributes {
+	return r.attrs
+}
+func (r *reader) As(i interface{}) bool {
+	p, ok := i.(*azblob.BlockBlobURL)
+	if !ok {
+		return false
+	}
+	*p = *r.raw
+	return true
+}
+
+// NewRangeReader implements driver.NewRangeReader.
+func (b *bucket) NewRangeReader(ctx context.Context, key string, offset, length int64, opts *driver.ReaderOptions) (driver.Reader, error) {
+
+	if key == "" {
+		return nil, fmt.Errorf("Invalid/Empty Key")
+	}
+
+	key = strings.Replace(key, OSPathSeparator, b.defaultDelimiter, -1)
+
+	blockBlobURL := b.urls.containerURL.NewBlockBlobURL(key)
+	blobPropertiesResponse, err := blockBlobURL.GetProperties(ctx, azblob.BlobAccessConditions{})
```

Comment:
Unfortunately, I have to do 2 requests to get the actual blob size (for tests). DownloadResponse’s contentLength is the http body size and not the blob content size. Tests such as ‘read a part in middle’ will fail because they expect the blob size, not length + offset.

Reply:
Yep, this works. Fixed in [changeset](https://github.com/google/go-cloud/pull/936/commits/8d301392023f41e43bb2d587a82dad02440e8582)